### ISSUE TITLE
nghttpx seems to delay the request to squid sometimes.

### ISSUE THREAD

**Issue by joeshae** [nghttp2/nghttp2#703]
nghttpx is used as a h2 proxy frontend, which backend is squid.
nghttpx and squid are running on the same dedicated server.
It's usually no problem when you browse the websites, but it takes long time for some requests occasionally. 
If you refresh the page when this issue happened, the issue disappears weirdly, even if you clean the cache before refreshing.
This occasional issue exists in both the latest stable Chrome and Firefox, so I don't think it's a browser's bug.

Let's take a look at an example in Firefox's Developer Tools:

![firefox](https://cloud.githubusercontent.com/assets/6749515/19221938/7938a17a-8e7f-11e6-98f0-fe22641433d3.png)

This is the log from nghttpx about this request:

![nghttpx](https://cloud.githubusercontent.com/assets/6749515/19221976/60fdd908-8e80-11e6-8f31-2e0a373c13e7.png)

This is the log from squid about this request:

![squid](https://cloud.githubusercontent.com/assets/6749515/19221974/49446f70-8e80-11e6-9329-fffb3be7f06f.png)

Some analysis:

1) Both nghttpx and squid won't print the log until a request is finished.
So from the same timestamps in nghttpx and squid, we can see, when squid returns the http response to nghttpx, **it takes less than 1 second for nghttpx to deliver it to the browser**. 
If the network was slow when nghttpx deliveried the response to the browser, the timestamp in nghttpx would be later than the timestamp in squid.

2) Take a look at the "145, -, 145, 145" in squid log.
Corresponding squid logformat is "%tr,%dt,%<pt,%<tt", which explains at:
http://www.squid-cache.org/Doc/config/logformat/

%tr	    Response time (milliseconds)
%dt	    Total time spent making DNS lookups (milliseconds)
%<pt    Peer response time in milliseconds. The timer starts when the last request byte is sent to the next hop and stops when the last response byte is received.
%<tt	    Total time in milliseconds. The timer starts with the first connect request (or write I/O) sent to the first selected peer. The timer stops with the last I/O with the last peer.

It means **it takes only 145 milliseconds for squid to finish the whole request**.

3) Take a look at the "15.180" in nghttpx log.
It means **it takes 15.180 seconds for nghttpx from receiving the request to finishing delivering the response to the client.**

So there is only one possibility:
nghttpx didn't transfer the raw request to squid at the beginning, which lasted for 15 seconds.

What's your opinion?

**Comment by tatsuhiro-t** [comment]
Your observation is correct.
What nghttpx does is that it limits the number of concurrent requests to the same origin host.
The limit is configurable by --backend-connections-per-host option, and it defaults to 8.  This is good number, because I heard that most browsers do 6 concurrent connections to the same host.
If nghttpx detects that there are already more than 8 backend requests for the same host, it blocks new request until old request is retired.  And 15 seconds measures from the time nghttpx received first byte of request.  With you observation, that means that request was blocked for ~15 seconds before it was sent to backend squid.  This may happen if the existing request takes long time, for example, its response header does not include content-length, and we have to wait EOF from backend.  I'm not sure this applies to your case.
Meanwhile, I could not reproduce this issue while browsing several sites.  I saw blocked requests, but they were quickly unblocked.
From the log, you are using spdy/3.1.  Are there any reason why you don't use h2 instead of spdy/3.1?

**Comment by tatsuhiro-t** [comment]
BTW, is this duplicate of #697?

**Comment by joeshae** [comment]
Even if I set backend-connections-per-host from 8 to 1024 and use h2, this issue appeared in Chrome and Firefox occasionally.

See the screenshots from Chome's Developer Tools:

![2-chrome](https://cloud.githubusercontent.com/assets/6749515/19226494/8704bd84-8ede-11e6-88c2-e3a149b3b4d6.png)

![2-chrome-b](https://cloud.githubusercontent.com/assets/6749515/19226497/8e431668-8ede-11e6-9a7f-94ad744ba92f.png)


this is log screenshot from nghttpx:

![2-nghttpx](https://cloud.githubusercontent.com/assets/6749515/19226512/a795c3b8-8ede-11e6-8cbd-26f5b7676783.png)


And this is log screenshot from squid:

![2-squid](https://cloud.githubusercontent.com/assets/6749515/19226507/9d46fe36-8ede-11e6-82ff-0f4a3e6972fb.png)



By the way, I'm not sure whether this issue is a duplicate of #697.
Because in this issue which ONLY blocks the related request itself without the whole browser's requests. But the issue in #697 would block the whole browser including all new and unfinished requests.

**Comment by tatsuhiro-t** [comment]
Thank you.  Could you provide us INFO level log of nghttpx, at least for around the slow requests?

**Comment by joeshae** [comment]
Sure. 
Because it's in production environment, I'll do it at night. 
Please wait for several hours.

**Comment by tatsuhiro-t** [comment]
Thanks.
One thing I'd like to ask is that rate limit options are used or not.

**Comment by joeshae** [comment]
No.
This is my nghttpx.conf:

```
daemon=yes
http2-proxy=yes
npn-list=spdy/3.1,h2,h2-16,h2-14,http/1.1

private-key-file=/usr/etc/cert/foobar.com.key
certificate-file=/usr/etc/cert/foobar.com.crt

frontend=0.0.0.0,443

backend=127.0.0.1,241
backend=127.0.0.1,242
backend=127.0.0.1,243
backend=127.0.0.1,244
backend=127.0.0.1,245
backend=127.0.0.1,246
backend=127.0.0.1,247
backend=127.0.0.1,248


workers=8
backlog=10240
fastopen=10240
server-name=nghx

backend-connections-per-host=1024

api-max-request-body=256K
backend-connections-per-host=1024
request-header-field-buffer=256K
max-request-header-fields=1000
response-header-field-buffer=256K
max-response-header-fields=1000

accesslog-file=/var/log/nghttpx/access-443.log
accesslog-format=[$time_local] $remote_addr:$remote_port->:$server_port $body_bytes_sent $request_time $status $pid $alpn $ssl_cipher $ssl_protocol $ssl_session_id $ssl_session_reused "$request" "$http_user_agent"
errorlog-file=/var/log/nghttpx/error-443.log
```

**Comment by joeshae** [comment]
When I set log-level to INFO, I can see this issue is caused by [#662](https://github.com/nghttp2/nghttp2/issues/662)

```
10/Oct/2016:19:15:43 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:185 [DCONN:0x7f7c05c1e700] Attaching to DOWNSTREAM:0x7f7c05c27a00
10/Oct/2016:19:15:43 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:278 [DCONN:0x7f7c05c1e700] Connecting to downstream server
10/Oct/2016:19:15:43 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:524 [DCONN:0x7f7c05c1e700] HTTP request headers. stream_id=1837

.......

10/Oct/2016:19:15:58 +0800 PID464 [WARN] shrpx_http_downstream_connection.cc:1167 [DCONN:0x7f7c05c1e700] Backend connect failed; addr=127.0.0.1:243: errno=110
10/Oct/2016:19:15:58 +0800 PID464 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.65189 seconds
10/Oct/2016:19:15:58 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:179 [DCONN:0x7f7c05c1e700] Deleted
10/Oct/2016:19:15:58 +0800 PID464 [INFO] shrpx_client_handler.cc:969 [CLIENT_HANDLER:0x7f7c05c45000] Downstream address group_idx: 0
10/Oct/2016:19:15:58 +0800 PID464 [INFO] shrpx_client_handler.cc:1077 [CLIENT_HANDLER:0x7f7c05c45000] Downstream connection pool is empty. Create new one
10/Oct/2016:19:15:58 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:185 [DCONN:0x7f7c05c1e700] Attaching to DOWNSTREAM:0x7f7c05c27a00
10/Oct/2016:19:15:58 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:278 [DCONN:0x7f7c05c1e700] Connecting to downstream server
10/Oct/2016:19:15:58 +0800 PID464 [INFO] shrpx_http_downstream_connection.cc:1177 [DCONN:0x7f7c05c1e700] Connected to downstream host
```

**It's obviously nghttpx waited for 15 seconds before it tried again with the warning "Backend connect failed", which was successful the second time.
Is it possible to set the timeout parameter to a lower value?**

This issue really has made me crazy for these months.
When squid started, it's no problem in the beginning few hours, then the issue appeared, more and more frequently. 
I've tried every effort I can to set the squid.conf in order to avoid this happed, but it's not work.
It seems squid doesn't provide a effective official support way.

**Comment by tatsuhiro-t** [comment]
Yeah, this is the exactly same issue with #662.
Currently, we don't have timeout for connect, but we use write timeout for it (--backend-write-timeout).
I'm fine to add option for specific to connect timeout.

I'm not familiar about squid development, but they have mailing list: http://www.squid-cache.org/Support/mailing-lists.html#squid-users

**Comment by tatsuhiro-t** [comment]
00a8c37 adds --backend-connect-timeout option.

**Comment by joeshae** [comment]
Is the default value of backend-connect-timeout 15 seconds?

**Comment by tatsuhiro-t** [comment]
No, 30 seconds.
I think the timeout in your log comes from TCP stack, not application level timeout.

**Comment by joeshae** [comment]
If the 15 seconds delay comes from TCP stack, do you know where I can adjust it? (CentOS)

**Comment by tatsuhiro-t** [comment]
No idea.

**Comment by joeshae** [comment]
Finally I got this:
https://blog.cloudflare.com/this-is-strictly-a-violation-of-the-tcp-specification/

This is what happend to me. 

Of course, this issue is not caused by nghttpx, but I think nghttpx can do something (control the source port when connecting to squid) to make it better, especially when working with some buggy softwares (squid etc.).

**Comment by tatsuhiro-t** [comment]
The blog post is very interesting read.  I agree that we have the same issue here.

I'm wondering how we can avoid this in our side.  Actually, we already choose client port in random fashion.  We don't perform "bind" in the backend connection.  Perhaps, short timeout with --backend-connect-timeout would workaround this a bit.

For the record, I found the similar issue in squid mailing list: http://lists.squid-cache.org/pipermail/squid-users/2015-November/007901.html

**Comment by joeshae** [comment]
On my server, squid works on local port 41~47.

When I see these logs in nghttpx.log:

```
12/Oct/2016:22:57:45 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f202e1c80] Connect time out; addr=127.0.0.1:47
12/Oct/2016:22:57:45 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.28845 seconds
12/Oct/2016:23:00:31 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f220cdd00] Connect time out; addr=127.0.0.1:47
12/Oct/2016:23:00:31 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.47325 seconds
12/Oct/2016:23:07:33 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f1f063180] Connect time out; addr=127.0.0.1:47
12/Oct/2016:23:07:33 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.91279 seconds
12/Oct/2016:22:57:45 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f202e1c80] Connect time out; addr=127.0.0.1:47
12/Oct/2016:22:57:45 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.28845 seconds
12/Oct/2016:23:00:31 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f220cdd00] Connect time out; addr=127.0.0.1:47
12/Oct/2016:23:00:31 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.47325 seconds
12/Oct/2016:23:07:33 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f1f063180] Connect time out; addr=127.0.0.1:47
12/Oct/2016:23:07:33 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.91279 seconds
12/Oct/2016:23:23:24 +0800 PID6900 [WARN] shrpx_http_downstream_connection.cc:75 [DCONN:0x7f6f20733a80] Connect time out; addr=127.0.0.1:47
12/Oct/2016:23:23:24 +0800 PID6900 [WARN] shrpx_connect_blocker.cc:93 Could not connect 1 times in a row; sleep for 1.53518 seconds
```

I ran this command to see there was a port pair (127.0.0.1:47, 127.0.0.1:52884), which exists for a very very long period:

```
[root@cn2-qz ~]# lsof -Pni:41 | grep CLOSE_WAIT; lsof -Pni:42 | grep CLOSE_WAIT; lsof -Pni:43 | grep CLOSE_WAIT; lsof -Pni:44 | grep CLOSE_WAIT; lsof -Pni:45 | grep CLOSE_WAIT; lsof -Pni:46 | grep CLOSE_WAIT; lsof -Pni:47 | grep CLOSE_WAIT; lsof -Pni:48 | grep CLOSE_WAIT;

nghttpx 6901   root  159u  IPv4 2401007      0t0  TCP 127.0.0.1:32981->127.0.0.1:46 (CLOSE_WAIT)
squid   6831 nobody  330u  IPv6 3697396      0t0  TCP 127.0.0.1:47->127.0.0.1:52884 (CLOSE_WAIT)
```

**Comment by joeshae** [comment]
I read this from your link:

```
> I read that connections that are in CLOSE_WAIT state don't have any tcp
> timeout, it's Squid that must close the socket.

Squid closes the socket/FD as soon as it received the FIN or RST that
began the CLOSE_WAIT state. Unless it was Squid closing that began it.
```

Do you think it is possible for nghttpx to send a FIN or RST to help squid terminate the CLOSE_WAIT state?

**Comment by tatsuhiro-t** [comment]
If CLOSE_WAIT on nghttpx appears very long time, it could be a problem.  We may stop reading from squid sometimes, and in that case, we don't immediately close the socket.  But this does not cause this issue, since nghttpx does not reuse such address.

> squid says it has CLOSE_WAIT states due to remote clients that go away without properly shutting down (either because of a network failure or because of a poor client implementation).

As far as I know, CLOSE_WAIT itself is OK.  It is a valid transition when local endpoint received connection termination request.

> Do you think it is possible for nghttpx to send a FIN or RST to help squid terminate the CLOSE_WAIT state?

Well, endpoint goes to CLOSE_WAIT because it received FIN from remote peer.
To transit from CLOSE_WAIT, endpoint has to close socket.

> Check the system's CLOSE_WAIT pairs before connecting, then prevent using the source ports which will be definitely timeout.

Don't know how to do this in standard socket API.
